warehouse/accounts/services.py

```diff
@@ -167,6 +180,50 @@ def verify_email(self, user_id, email_address):
                 email.verified = True
 
 
+@implementer(IUserTokenService)
+class UserTokenService:
+    def __init__(self, user_service, settings):
+        self.user_service = user_service
+        self.serializer = URLSafeTimedSerializer(
+            settings["password_reset.secret"],
+            salt="password-reset",
+        )
+        self.token_max_age = settings["password_reset.token_max_age"]
+
+    def generate_token(self, user):
+        return self.serializer.dumps({
+            "user.id": str(user.id),
+            "user.last_login": str(user.last_login),
+            "user.password_date": str(user.password_date),
+        })
+
+    def get_user_by_token(self, token):
+        if not token:
+            raise InvalidPasswordResetToken("Invalid token")
+
+        try:
+            data = self.serializer.loads(token, max_age=self.token_max_age)
+        except SignatureExpired:
+            raise InvalidPasswordResetToken("Expired token")
+        except BadData: #  Catch all other exceptions
+            raise InvalidPasswordResetToken("Invalid token")
+
```

Comment:
Could we make these error messages a bit more helpful to the user? e.g.

"This reset token has expired -  please request a new link"
"This reset token is invalid - please request a new link"